File: umbra-typescript/src/device.ts
import type { ApiClient } from './api'
import type { DeviceCredential } from './store'
import type { DeviceCredentialStore } from './store'
import { isDeviceSessionClosedError, UmbraError } from './errors'

export const registrationDeviceId = 'registration'
const registrationTokenPrefix = 'umbra_reg_v1_'

export interface DeviceMetadata {
  name: string
  platform?: string
  os_version?: string
  app_version?: string
  fingerprint?: string
  metadata?: Record<string, unknown>
  readonly __umbraAutoCollectedDeviceMetadata: true
}

export interface DeviceRegistrationInput {
  device: DeviceMetadata
  credentialId?: string
  credentialSecret?: string
  registrationToken?: string
}

interface DeviceRegistrationRequest {
  credential_id?: string
  registration_token?: string
  device: DeviceMetadata
}

export type CollectedDeviceMetadataFields = Omit<DeviceMetadata, '__umbraAutoCollectedDeviceMetadata'>

const autoCollectedDeviceMetadata = new WeakSet<object>()

export function markAutoCollectedDeviceMetadata(metadata: CollectedDeviceMetadataFields): DeviceMetadata {
  autoCollectedDeviceMetadata.add(metadata)
  Object.defineProperty(metadata, '__umbraAutoCollectedDeviceMetadata', {
    value: true,
    enumerable: false,
    configurable: false,
  })
  return metadata as DeviceMetadata
}

function isAutoCollectedDeviceMetadata(metadata: DeviceMetadata) {
  return typeof metadata === 'object' && metadata !== null && autoCollectedDeviceMetadata.has(metadata)
}

export interface ClientDevice {
  id?: number
  device_id: string
  user_id?: number
  tenant_id?: number
  client_id?: string
  distribution_credential_id?: number
  distribution_credential_key?: string
  name?: string
  platform?: string
  os_version?: string
  app_version?: string
  fingerprint?: string
  metadata?: Record<string, unknown>
  status?: number
  created_at?: string
  updated_at?: string
  last_active_at?: string
  rotated_at?: string
  revoked_at?: string
}

export interface DeviceRegistrationResult {
  device: ClientDevice
  device_secret: string
  secret_once: boolean
}

export interface DeviceSignatureInput {
  method: string
  pathWithQuery: string
  timestamp: number
  nonce: string
  body: string
  deviceId: string
  secret: string
}

export interface DeviceSignatureResult {
  canonicalString: string
  bodyHash: string
  signature: string
}

export interface DeviceSignatureHeadersInput {
  method: string
  url: URL
  body: string
  credential: DeviceCredential
  includeDeviceId?: boolean
}

export class DeviceClient {
  readonly #api: ApiClient
  readonly #store: DeviceCredentialStore
  readonly #defaultRegistration: DeviceRegistrationInput | undefined

  constructor(api: ApiClient, store: DeviceCredentialStore, defaultRegistration?: DeviceRegistrationInput) {
    this.#api = api
    this.#store = store
    this.#defaultRegistration = defaultRegistration
  }

  async register(input: DeviceRegistrationInput) {
    const credential = resolveRegistrationCredential(input)
    if (!isAutoCollectedDeviceMetadata(input.device)) {
      throw UmbraError.invalidInput('device metadata must be collected by the SDK')
    }
    const credentialId = input.credentialId?.trim() || ''
    const request: DeviceRegistrationRequest = {
      device: input.device,
      ...(credentialId ? { credential_id: credentialId } : {}),
      ...(input.registrationToken ? { registration_token: input.registrationToken } : {}),
    }
    const response = await this.#api.registerDevice<DeviceRegistrationResult>(request, credential.secret)
    if (!response.device.device_id || !response.device_secret) {
      throw UmbraError.invalidInput('device registration response is missing credentials')
    }
    await this.#store.save({
      deviceId: response.device.device_id,
      deviceSecret: response.device_secret,
    })
    return response
  }

  async ensureRegistered(input: DeviceRegistrationInput) {
    const credential = await this.#store.load()
    if (credential?.deviceId && credential.deviceSecret) return credential
    const response = await this.register(input)
    return {
      deviceId: response.device.device_id,
      deviceSecret: response.device_secret,
    }
  }

  async registerDefault() {
    if (!this.#defaultRegistration) {
      throw UmbraError.invalidInput('device registration is not configured')
    }
    return this.register(this.#defaultRegistration)
  }

  async ensureDefaultRegistered() {
    if (!this.#defaultRegistration) {
      throw UmbraError.invalidInput('device registration is not configured')
    }
    return this.ensureRegistered(this.#defaultRegistration)
  }

  async logout() {
    const credential = await this.#store.load()
    let reportError: unknown
    try {
      if (credential?.deviceId && credential.deviceSecret) {
        await this.#api.post<ClientDevice>('/client/devices/logout', {})
      }
    }
    catch (error) {
      if (!isDeviceSessionClosedError(error)) reportError = error
    }
    let clearError: unknown
    try {
      await this.#store.clear()
    }
    catch (error) {
      clearError = error
    }
    if (reportError && clearError) throw new AggregateError([reportError, clearError], 'device logout failed')
    if (reportError) throw reportError
    if (clearError) throw clearError
  }

  async rotateSecret(deviceId?: string) {
    const stored = await this.#store.load()
    const targetDeviceId = deviceId?.trim() || stored?.deviceId || ''
    if (!targetDeviceId) {
      throw UmbraError.invalidInput('deviceId is required')
    }
    const response = await this.#api.post<DeviceRegistrationResult>(
      `/user/devices/${encodeURIComponent(targetDeviceId)}/rotate-secret`,
      {},
    )
    if (!response.device.device_id || !response.device_secret) {
      throw UmbraError.invalidInput('device secret rotation response is missing credentials')
    }
    await this.#store.save({
      deviceId: response.device.device_id,
      deviceSecret: response.device_secret,
    })
    return response
  }

  loadCredential() {
    return this.#store.load()
  }

  saveCredential(credential: DeviceCredential) {
    return this.#store.save(credential)
  }

  clearCredential() {
    return this.#store.clear()
  }
}

export async function createDeviceSignature(input: DeviceSignatureInput): Promise<DeviceSignatureResult> {
  const bodyHash = await sha256Base64Url(input.body)
  const canonicalString = buildDeviceCanonicalString({
    method: input.method,
    pathWithQuery: input.pathWithQuery,
    timestamp: input.timestamp,
    nonce: input.nonce,
    bodyHash,
    deviceId: input.deviceId,
  })
  const signature = await hmacSha256Base64Url(input.secret, canonicalString)
  return { canonicalString, bodyHash, signature }
}

export function buildDeviceCanonicalString(input: {
  method: string
  pathWithQuery: string
  timestamp: number
  nonce: string
  bodyHash: string
  deviceId: string
}) {
  return [
    'v1',
    input.method.toUpperCase(),
    input.pathWithQuery,
    String(input.timestamp),
    input.nonce,
    input.bodyHash,
    input.deviceId,
  ].join('\n')
}

export async function createDeviceSignatureHeaders(input: DeviceSignatureHeadersInput) {
  const timestamp = Math.floor(Date.now() / 1000)
  const nonce = randomNonce()
  const pathWithQuery = `${input.url.pathname}${input.url.search}`
  const signature = await createDeviceSignature({
    method: input.method,
    pathWithQuery,
    timestamp,
    nonce,
    body: input.body,
    deviceId: input.credential.deviceId,
    secret: input.credential.deviceSecret,
  })

  return {
    ...(input.includeDeviceId === false ? {} : { 'X-Umbra-Device-Id': input.credential.deviceId }),
    'X-Umbra-Timestamp': String(timestamp),
    'X-Umbra-Nonce': nonce,
    'X-Umbra-Body-SHA256': signature.bodyHash,
    'X-Umbra-Signature': `v1=${signature.signature}`,
  }
}

export function parseRegistrationToken(token: string) {
  const trimmed = token.trim()
  if (!trimmed.startsWith(registrationTokenPrefix)) {
    throw UmbraError.invalidInput('registrationToken must start with umbra_reg_v1_')
  }
  const rest = trimmed.slice(registrationTokenPrefix.length)
  const dot = rest.indexOf('.')
  if (dot <= 0 || dot === rest.length - 1) {
    throw UmbraError.invalidInput('registrationToken must include credential id and secret')
  }
  const secret = rest.slice(dot + 1)
  if (secret.includes('.')) {
    throw UmbraError.invalidInput('registrationToken secret must not contain dots')
  }
  return {
    credentialId: rest.slice(0, dot),
    secret,
  }
}

export async function sha256Base64Url(body: string) {
  const digest = await cryptoSubtle().digest('SHA-256', new TextEncoder().encode(body))
  return base64Url(new Uint8Array(digest))
}

function resolveRegistrationCredential(input: DeviceRegistrationInput) {
  const tokenCredential = input.registrationToken ? parseRegistrationToken(input.registrationToken) : null
  const credentialId = input.credentialId?.trim() || tokenCredential?.credentialId || ''
  if (!credentialId) {
    throw UmbraError.invalidInput('credentialId or registrationToken is required')
  }
  if (tokenCredential && input.credentialId?.trim() && input.credentialId.trim() !== tokenCredential.credentialId) {
    throw UmbraError.invalidInput('credentialId does not match registrationToken')
  }
  const secret = input.credentialSecret?.trim() || tokenCredential?.secret || ''
  if (!secret) {
    throw UmbraError.invalidInput('credentialSecret or registrationToken is required')
  }
  return { credentialId, secret }
}

async function hmacSha256Base64Url(secret: string, message: string) {
  const subtle = cryptoSubtle()
  const key = await subtle.importKey(
    'raw',
    new TextEncoder().encode(secret),
    { name: 'HMAC', hash: 'SHA-256' },
    false,
    ['sign'],
  )
  const signature = await subtle.sign('HMAC', key, new TextEncoder().encode(message))
  return base64Url(new Uint8Array(signature))
}

function randomNonce() {
  return base64Url(randomBytes(24))
}

function randomBytes(size: number) {
  const crypto = cryptoObject()
  const bytes = new Uint8Array(size)
  crypto.getRandomValues(bytes)
  return bytes
}

function cryptoObject() {
  if (!globalThis.crypto?.getRandomValues) {
    throw UmbraError.invalidInput('Web Crypto getRandomValues is not available')
  }
  return globalThis.crypto
}

function cryptoSubtle() {
  if (!globalThis.crypto?.subtle) {
    throw UmbraError.invalidInput('Web Crypto subtle digest is not available')
  }
  return globalThis.crypto.subtle
}

function base64Url(bytes: Uint8Array) {
  let base64: string
  if (typeof btoa === 'function') {
    let binary = ''
    for (const byte of bytes) binary += String.fromCharCode(byte)
    base64 = btoa(binary)
  }
  else {
    base64 = Buffer.from(bytes).toString('base64')
  }
  return base64.replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/g, '')
}

File: umbra-go/device.go
package umbra

import (
	"context"
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"io"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"time"
)

const (
	deviceSignatureVersion  = "v1"
	registrationTokenPrefix = "umbra_reg_v1_"
	registrationDeviceID    = "registration"

	headerDeviceID   = "X-Umbra-Device-Id"
	headerTimestamp  = "X-Umbra-Timestamp"
	headerNonce      = "X-Umbra-Nonce"
	headerBodySHA256 = "X-Umbra-Body-SHA256"
	headerSignature  = "X-Umbra-Signature"
)

type DeviceClient struct {
	api                 *apiClient
	store               DeviceStore
	defaultRegistration *DeviceRegistrationOptions
}

type DeviceCredentials struct {
	DeviceID     string `json:"device_id"`
	DeviceSecret string `json:"device_secret"`
}

type DeviceMetadata struct {
	Name        string         `json:"name"`
	Platform    string         `json:"platform,omitempty"`
	AppVersion  string         `json:"app_version,omitempty"`
	OSVersion   string         `json:"os_version,omitempty"`
	Fingerprint string         `json:"fingerprint,omitempty"`
	Metadata    map[string]any `json:"metadata,omitempty"`

	autoCollected bool
}

type DeviceRegistrationOptions struct {
	CredentialID      string
	CredentialSecret  string
	RegistrationToken string
	Device            DeviceMetadata
}

type Device struct {
	ID                        uint64         `json:"id"`
	DeviceID                  string         `json:"device_id"`
	UserID                    uint64         `json:"user_id"`
	TenantID                  *uint64        `json:"tenant_id,omitempty"`
	ClientID                  string         `json:"client_id"`
	DistributionCredentialID  uint64         `json:"distribution_credential_id"`
	DistributionCredentialKey string         `json:"distribution_credential_key,omitempty"`
	Name                      string         `json:"name"`
	Platform                  string         `json:"platform"`
	AppVersion                string         `json:"app_version"`
	OSVersion                 string         `json:"os_version"`
	Fingerprint               string         `json:"fingerprint,omitempty"`
	Metadata                  map[string]any `json:"metadata,omitempty"`
	Status                    uint8          `json:"status"`
	CreatedAt                 time.Time      `json:"created_at"`
	UpdatedAt                 time.Time      `json:"updated_at"`
	LastActiveAt              *time.Time     `json:"last_active_at,omitempty"`
	RotatedAt                 *time.Time     `json:"rotated_at,omitempty"`
	RevokedAt                 *time.Time     `json:"revoked_at,omitempty"`
}

type DeviceRegistrationResult struct {
	Device       Device `json:"device"`
	DeviceSecret string `json:"device_secret,omitempty"`
	SecretOnce   bool   `json:"secret_once"`
}

type deviceRegistrationRequest struct {
	CredentialID      string         `json:"credential_id,omitempty"`
	RegistrationToken string         `json:"registration_token,omitempty"`
	Device            DeviceMetadata `json:"device"`
}

// Register registers the current authenticated user device and stores the
// returned one-time device_secret.
func (d *DeviceClient) Register(ctx context.Context, options DeviceRegistrationOptions) (*DeviceRegistrationResult, error) {
	credentialID, credentialSecret, err := resolveRegistrationSecret(options)
	if err != nil {
		return nil, err
	}
	if strings.TrimSpace(options.Device.Name) == "" {
		return nil, invalidInput("device name is required")
	}
	if !options.Device.autoCollected {
		return nil, invalidInput("device metadata must be collected by the SDK")
	}
	requestCredentialID := strings.TrimSpace(options.CredentialID)
	if strings.TrimSpace(options.RegistrationToken) == "" {
		requestCredentialID = credentialID
	}
	body := deviceRegistrationRequest{
		CredentialID:      requestCredentialID,
		RegistrationToken: strings.TrimSpace(options.RegistrationToken),
		Device:            options.Device,
	}
	var out DeviceRegistrationResult
	if err := d.api.postSigned(ctx, "/client/devices/register", body, &out, requestSigner{
		Secret:   credentialSecret,
		DeviceID: registrationDeviceID,
	}); err != nil {
		return nil, err
	}
	if out.Device.DeviceID == "" || out.DeviceSecret == "" {
		return nil, invalidInput("device registration response missing credentials")
	}
	if d.store != nil {
		if err := d.store.Save(ctx, &DeviceCredentials{DeviceID: out.Device.DeviceID, DeviceSecret: out.DeviceSecret}); err != nil {
			return nil, err
		}
	}
	return &out, nil
}

// EnsureRegistered returns stored device credentials or registers the device
// using the supplied options when no credentials are stored.
func (d *DeviceClient) EnsureRegistered(ctx context.Context, options DeviceRegistrationOptions) (*DeviceCredentials, error) {
	if d.store != nil {
		credentials, err := d.store.Load(ctx)
		if err != nil {
			return nil, err
		}
		if credentials != nil && credentials.DeviceID != "" && credentials.DeviceSecret != "" {
			return credentials, nil
		}
	}
	result, err := d.Register(ctx, options)
	if err != nil {
		return nil, err
	}
	return &DeviceCredentials{DeviceID: result.Device.DeviceID, DeviceSecret: result.DeviceSecret}, nil
}

// RegisterDefault registers with Config.DeviceRegistration.
func (d *DeviceClient) RegisterDefault(ctx context.Context) (*DeviceRegistrationResult, error) {
	if d.defaultRegistration == nil {
		return nil, invalidInput("device registration is not configured")
	}
	return d.Register(ctx, *d.defaultRegistration)
}

// EnsureDefaultRegistered ensures a device is registered with Config.DeviceRegistration.
func (d *DeviceClient) EnsureDefaultRegistered(ctx context.Context) (*DeviceCredentials, error) {
	if d.defaultRegistration == nil {
		return nil, invalidInput("device registration is not configured")
	}
	return d.EnsureRegistered(ctx, *d.defaultRegistration)
}

// RotateSecret rotates the current user's device secret and stores the new
// secret when the rotated device matches the local credentials.
func (d *DeviceClient) RotateSecret(ctx context.Context, deviceID string) (*DeviceRegistrationResult, error) {
	deviceID = strings.TrimSpace(deviceID)
	if deviceID == "" && d.store != nil {
		credentials, err := d.store.Load(ctx)
		if err != nil {
			return nil, err
		}
		if credentials != nil {
			deviceID = credentials.DeviceID
		}
	}
	if deviceID == "" {
		return nil, invalidInput("device id is required")
	}
	var out DeviceRegistrationResult
	if err := d.api.post(ctx, "/user/devices/"+url.PathEscape(deviceID)+"/rotate-secret", map[string]any{}, &out); err != nil {
		return nil, err
	}
	if out.Device.DeviceID == "" || out.DeviceSecret == "" {
		return nil, invalidInput("device secret rotation response missing credentials")
	}
	if d.store != nil {
		if err := d.store.Save(ctx, &DeviceCredentials{DeviceID: out.Device.DeviceID, DeviceSecret: out.DeviceSecret}); err != nil {
			return nil, err
		}
	}
	return &out, nil
}

func resolveRegistrationSecret(options DeviceRegistrationOptions) (string, string, error) {
	credentialID := strings.TrimSpace(options.CredentialID)
	credentialSecret := strings.TrimSpace(options.CredentialSecret)
	registrationToken := strings.TrimSpace(options.RegistrationToken)
	if registrationToken != "" {
		tokenCredentialID, tokenSecret, err := ParseRegistrationToken(registrationToken)
		if err != nil {
			return "", "", err
		}
		if credentialID != "" && credentialID != tokenCredentialID {
			return "", "", invalidInput("credential id does not match registration token")
		}
		credentialID = tokenCredentialID
		credentialSecret = tokenSecret
	}
	if credentialID == "" || credentialSecret == "" {
		return "", "", invalidInput("registration credential id and secret are required")
	}
	return credentialID, credentialSecret, nil
}

func MakeRegistrationToken(credentialID, credentialSecret string) (string, error) {
	credentialID = strings.TrimSpace(credentialID)
	credentialSecret = strings.TrimSpace(credentialSecret)
	if credentialID == "" || credentialSecret == "" || strings.Contains(credentialID, ".") || strings.Contains(credentialSecret, ".") {
		return "", invalidInput("registration credential id and secret are required")
	}
	return registrationTokenPrefix + credentialID + "." + credentialSecret, nil
}

func ParseRegistrationToken(token string) (string, string, error) {
	token = strings.TrimSpace(token)
	if !strings.HasPrefix(token, registrationTokenPrefix) {
		return "", "", invalidInput("registration token is invalid")
	}
	payload := strings.TrimPrefix(token, registrationTokenPrefix)
	credentialID, credentialSecret, ok := strings.Cut(payload, ".")
	credentialID = strings.TrimSpace(credentialID)
	credentialSecret = strings.TrimSpace(credentialSecret)
	if !ok || credentialID == "" || credentialSecret == "" || strings.Contains(credentialSecret, ".") {
		return "", "", invalidInput("registration token is invalid")
	}
	return credentialID, credentialSecret, nil
}

type requestSigner struct {
	Secret           string
	DeviceID         string
	SkipDeviceHeader bool
}

func signRequest(req *http.Request, body []byte, signer requestSigner) error {
	secret := strings.TrimSpace(signer.Secret)
	deviceID := strings.TrimSpace(signer.DeviceID)
	if secret == "" || deviceID == "" {
		return invalidInput("device signing credentials are required")
	}
	timestamp := time.Now().Unix()
	nonce, err := newNonce()
	if err != nil {
		return err
	}
	bodyHash := BodySHA256Base64URL(body)
	canonical := CanonicalString(req.Method, canonicalPathWithQuery(req.URL), timestamp, nonce, bodyHash, deviceID)
	signature := SignDeviceRequest(secret, canonical)
	if !signer.SkipDeviceHeader && deviceID != registrationDeviceID {
		req.Header.Set(headerDeviceID, deviceID)
	}
	req.Header.Set(headerTimestamp, strconv.FormatInt(timestamp, 10))
	req.Header.Set(headerNonce, nonce)
	req.Header.Set(headerBodySHA256, bodyHash)
	req.Header.Set(headerSignature, deviceSignatureVersion+"="+signature)
	return nil
}

func CanonicalString(method, pathWithQuery string, timestamp int64, nonce, bodyHash, deviceID string) string {
	return strings.Join([]string{
		deviceSignatureVersion,
		strings.ToUpper(strings.TrimSpace(method)),
		pathWithQuery,
		strconv.FormatInt(timestamp, 10),
		strings.TrimSpace(nonce),
		strings.TrimSpace(bodyHash),
		strings.TrimSpace(deviceID),
	}, "\n")
}

func BodySHA256Base64URL(body []byte) string {
	sum := sha256.Sum256(body)
	return base64.RawURLEncoding.EncodeToString(sum[:])
}

func SignDeviceRequest(secret string, canonical string) string {
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(canonical))
	return base64.RawURLEncoding.EncodeToString(mac.Sum(nil))
}

func canonicalPathWithQuery(u *url.URL) string {
	if u == nil {
		return "/"
	}
	path := u.EscapedPath()
	if path == "" {
		path = "/"
	}
	if u.RawQuery != "" {
		return path + "?" + u.RawQuery
	}
	return path
}

func newNonce() (string, error) {
	buf := make([]byte, 18)
	if _, err := io.ReadFull(rand.Reader, buf); err != nil {
		return "", err
	}
	return base64.RawURLEncoding.EncodeToString(buf), nil
}
